### PR TITLE
Use matrix exp for linear systems in dsolve

### LINKED ISSUE #14312: Incorrect solution of 3 by 3 linear ODE systems
`dsolve` gives wrong solutions of the following 3 by 3 homogeneous linear system. Specifically, the first equation is the one not being satisfied by the solution. If the equations are reordered, the solution changes so that again, 2nd and 3rd equations are satisfied but the 1st is not. [From Stack Overflow](https://stackoverflow.com/questions/48949289/sympy-dsolve-answers-depends-on-the-order-of-the-input)
```
k2, k3, t = symbols('k2 k3 t')
x, y, z = symbols('x y z', cls=Function)
eq1 = (Eq(Derivative(x(t),t), k3*y(t)), Eq(Derivative(y(t),t), -(k3+k2)*y(t)), Eq(Derivative(z(t),t), k2*y(t)))
sol1 = dsolve(eq1)
print([eqn.subs({e.lhs: e.rhs for e in sol1}).doit().simplify() for eqn in eq1])
eq2 = (Eq(Derivative(z(t),t), k2*y(t)), Eq(Derivative(x(t),t), k3*y(t)), Eq(Derivative(y(t),t), -(k3+k2)*y(t)))
sol2 = dsolve(eq2)
print([eqn.subs({e.lhs: e.rhs for e in sol2}).doit().simplify() for eqn in eq2])
```
Expected output: [True, True, True] for both solutions. Actual output from current master: 
```
[Eq(-C3*k3*(k2 + k3)*exp(-t*(k2 + k3))/k2, (C2*k2*exp(t*(k2 + k3)) - C3*k3*(k2 + k3))*exp(-t*(k2 + k3))/k2), True, True]
```
and
```
[Eq(C3*k2*exp(-t*(k2 + k3)), C2 + C3*k2*exp(-k2*t - k3*t)), True, True]
```

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### References to other Issues or PRs

Fixes #14312

See also #15474 and #15407 

#### Brief description of what is fixed or changed

Add matrix exponential `matrix_exp(A, t)` to compute `exp(A t)` for a matrix `A`. Use this to solve first order systems of linear homogeneous constant coefficient ODEs with `dsolve`.

The existing code for this only works for systems whose matrix has a particular structure of Jordan blocks (discussed in #14312. This version uses the `jordan_form` to fully form the matrix exponential in all cases.

Using the matrix exponential directly results in overly complicated expressions in the case of symbolic coefficients. Given `exp(A t)` the solution for any initial value problemis `exp(A t) x0` where `x0` is the vector of initial conditions. Since we only compute the general solution here `x0` is an arbitrary constant vector `c` so we would naturally return `exp(A t) c`.

This method computes the `exp(A t)` through it's Jordan normal form as `P * J * P.inv()`. Evaluating that directly and using it for the general solution leads to complicated expressions and fails the only existing test for `linear_neq_order1_type1`. Instead I use the fact that:
```
     exp(A t) c = P * J * P.inv() * c = P * J * c'
```
where `c'` is a different arbitrary constant vector. Further down the line this might make it more complicated for the initial condition handling code. It gives solutions that are as close as possible to the current behaviour though and also results in simpler general solutions. It also means we don't need to invert `P` and is efficient.

Complex conjugate eigenvalue pairs are handled using the "real Jordan form" as described
[here on Wikipedia](https://en.wikipedia.org/wiki/Jordan_normal_form#Real_matrices). This allows us to naturally get solutions in terms of sin/cos rather than exp where appropriate. This is only done for real matrices: if `A.has(I)` then no sin/cos rewriting is done.

#### Other comments

This PR should mean that `linear_neq_order1_type1` can theoretically handle any system of linear, constant coefficient, homogeneous ODEs. There are a number of other system solvers such as `linear_2eq_order1_type1` that are special cases of this. If this solver can be made good enough then those others should probably be removed.

The matrix exponential can also be used to solve the corresponding non-homogeneous problems [as described here](https://en.wikipedia.org/wiki/Matrix_exponential#Inhomogeneous_case_generalization:_variation_of_parameters). It is also possible to rewrite any higher-order linear, constant coefficient ODE as a first order system.

With these things together this solver could solve any system of linear, constant coefficient ODEs homogeneous or non-homogeneous of any order (including mixed). That would be a big improvement in `dsolve` for systems which otherwise only handles a few special cases each involving a fixed number of equations.

#### Release Notes

* solvers
    * Use matrix exponential to solve homogeneous systems of first order ODEs with constant coefficients.

### CODE REVIEW

**Comment by snusstofilen** [comment]
Tried this version for the following equation and it didn't work

```
Eq(-b1(t) - v1(t)/r, c*Derivative(v1(t), t))
Eq(-r*b1(t) + v1(t), l*Derivative(b1(t), t))
```

For the initial conditions ``` {b1.subs({t:0}):0.096, v1.subs({t:0}):4} ```.

**Comment by oscarbenjamin** [comment]
This PR changes `_linear_neq_order1_type1` which is used for systems of 4 or more 1st order, linear, constant coefficient ODEs. There are specific solvers for 2 and for 3 equation systems. Probably this neq solver should be used for all of them though. To test the solver here use:
```python
from sympy import *
b1, v1 = symbols('b1, v1', cls=Function)
d1, d2 = symbols('d1, d2', cls=Function) # Dummy variables...
t, r, l, c = symbols('t, r, l, c')
eq = [
    Eq(-b1(t) - v1(t)/r, c*Derivative(v1(t), t)),
    Eq(-r*b1(t) + v1(t), l*Derivative(b1(t), t)),
    # Dummy equations to make it 4x4:
    Eq(d1(t).diff(t), 0),
    Eq(d2(t).diff(t), 0),
]
pprint(dsolve(eq, [b1(t), v1(t), d1(t), d2(t)],
                   ics={b1(0):0.096, v1(0):4, d1(0):0, d2(0):0}))
```
That gives:
```

```⎡                                                                                                     ⎛   _______________________________         ⎞                                                              
⎢                                                                                                     ⎜  ╱ ⎛ 2          ⎞ ⎛ 2          ⎞     2    ⎟                                                              
⎢                                                                                                     ⎜╲╱  ⎝r  - 2⋅r - 1⎠⋅⎝r  + 2⋅r - 1⎠    r  + 1⎟                                                              
⎢              ⎛        _______________    ⎞ ⎛                          ___________________      ⎞  t⋅⎜────────────────────────────────── + ──────⎟         ⎛        _______________    ⎞ ⎛                      
⎢              ⎜ 2     ╱  4      2         ⎟ ⎜     2                   ╱  4        2             ⎟    ⎝               2⋅r                    2⋅r  ⎠         ⎜ 2     ╱  4      2         ⎟ ⎜       2              
⎢        0.008⋅⎝r  - ╲╱  r  - 6⋅r  + 1  - 1⎠⋅⎝3.0⋅r  - 250.0⋅r + 3.0⋅╲╱  r  - 6.0⋅r  + 1.0  - 3.0⎠⋅ℯ                                                  0.008⋅⎝r  + ╲╱  r  - 6⋅r  + 1  - 1⎠⋅⎝- 3.0⋅r  + 250.0⋅r + 3
⎢v₁(t) = ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── + ───────────────────────────────────────────────────────────
⎢                                                                      ___________________                                                                                                                       
⎢                                                                     ╱  4        2                                                                                                                              
⎣                                                                 r⋅╲╱  r  - 6.0⋅r  + 1.0                                                                                                                        

                                    ⎛     _______________________________         ⎞                                                                           ⎛     _______________________________         ⎞    
                                    ⎜    ╱ ⎛ 2          ⎞ ⎛ 2          ⎞     2    ⎟                                                                           ⎜    ╱ ⎛ 2          ⎞ ⎛ 2          ⎞     2    ⎟    
                                    ⎜  ╲╱  ⎝r  - 2⋅r - 1⎠⋅⎝r  + 2⋅r - 1⎠    r  + 1⎟                                                                           ⎜  ╲╱  ⎝r  - 2⋅r - 1⎠⋅⎝r  + 2⋅r - 1⎠    r  + 1⎟    
      ___________________      ⎞  t⋅⎜- ────────────────────────────────── + ──────⎟                ⎛                            ___________________      ⎞  t⋅⎜- ────────────────────────────────── + ──────⎟    
     ╱  4        2             ⎟    ⎝                 2⋅r                    2⋅r  ⎠                ⎜       2                   ╱  4        2             ⎟    ⎝                 2⋅r                    2⋅r  ⎠    
.0⋅╲╱  r  - 6.0⋅r  + 1.0  + 3.0⎠⋅ℯ                                                           0.016⋅⎝- 3.0⋅r  + 250.0⋅r + 3.0⋅╲╱  r  - 6.0⋅r  + 1.0  + 3.0⎠⋅ℯ                                                    0
───────────────────────────────────────────────────────────────────────────────────, b₁(t) = ──────────────────────────────────────────────────────────────────────────────────────────────────────────────── + ─
     ___________________                                                                                                                     ___________________                                                 
    ╱  4        2                                                                                                                           ╱  4        2                                                        
r⋅╲╱  r  - 6.0⋅r  + 1.0                                                                                                                   ╲╱  r  - 6.0⋅r  + 1.0                                                  

                                                              ⎛   _______________________________         ⎞                          ⎤
                                                              ⎜  ╱ ⎛ 2          ⎞ ⎛ 2          ⎞     2    ⎟                          ⎥
                                                              ⎜╲╱  ⎝r  - 2⋅r - 1⎠⋅⎝r  + 2⋅r - 1⎠    r  + 1⎟                          ⎥
     ⎛                          ___________________      ⎞  t⋅⎜────────────────────────────────── + ──────⎟                          ⎥
     ⎜     2                   ╱  4        2             ⎟    ⎝               2⋅r                    2⋅r  ⎠                          ⎥
.016⋅⎝3.0⋅r  - 250.0⋅r + 3.0⋅╲╱  r  - 6.0⋅r  + 1.0  - 3.0⎠⋅ℯ                                                                         ⎥
───────────────────────────────────────────────────────────────────────────────────────────────────────────, d₁(t) = 0.0, d₂(t) = 0.0⎥
                                             ___________________                                                                     ⎥
                                            ╱  4        2                                                                            ⎥
                                          ╲╱  r  - 6.0⋅r  + 1.0                                                                      ⎦

```
Without the dummy equations the 2eq solver will be used and that gives the annoying piecewise solutions.

**Comment by czgdp1807** [comment]
Does this PR still holds a chance to be merged?
@cbm755 Please review this PR. @oscarbenjamin Please resolve the conflicts. Thanks.

File: sympy/solvers/ode.py [hunk 1]
```diff
@@ -7982,6 +7982,120 @@ def sysode_linear_neq_order1(match_):
...
+    differential equations. For example if x is a vector and A is a matrix
+    then the initial value problem
+
+        dx(t)/dt = A x(t),   x(0) = x0
+
+    has the unique solution
+
+        x(t) = exp(A t) x0
+
+    Example:
```
Comment (czgdp1807) [inline sympy/solvers/ode.py]: Probably, 
```
Example
=======
```

File: sympy/solvers/ode.py [hunk 2]
```diff
@@ -7982,6 +7982,120 @@ def sysode_linear_neq_order1(match_):
...
+    [      2*e  *sin(t)         - 3*e  *sin(t) + e  *cos(t)]
+    '''
+    P, J = matrix_exp_jordan_form(A, t)
+    return P * J * P.inv()
+
+def matrix_exp_jordan_form(A, t):
+    '''Matrix exponential exp(A*t) for the matrix A and scalar t.
+
+    Returns the Jordan form of the exp(A t).
+
```
Comment (czgdp1807) [inline sympy/solvers/ode.py]:
```
Example
=======
```

**Comment by oscarbenjamin** [comment]
This PR needs a bit of a rewrite. At the time I didn't know about the existing code for calculating matrix exponentials. Probably that should be adapted for use here.

Also the class of ODEs that this solves generalises many of the cases covered by special case solvers 2eq and 3eq solvers. It should be changed so that we have a single neq solver for this class of equations.

**Comment by mijo2** [comment]
I am going to revive this PR if that is ok?

**Comment by oscarbenjamin** [comment]
Yes, that's fine.

**Comment by mijo2** [comment]
Ok I will revive this.

**Comment by mijo2** [comment]
@oscarbenjamin I am reviving this PR by moving the ode.py and test_ode.py from solvers to solvers/ode.

File: sympy/solvers/tests/test_ode.py [hunk 3]
```diff
@@ -2934,8 +2936,207 @@ def test_issue_14395():
...
+
+    A = Matrix([[2, -5], [2, -4]])
+    expAt = Matrix([
+            [3*exp(-t)*sin(t) + exp(-t)*cos(t), -5*exp(-t)*sin(t)],
+            [2*exp(-t)*sin(t), -3*exp(-t)*sin(t) + exp(-t)*cos(t)]
+            ])
+    assert matrix_exp(A, t) == expAt
+
+    A = Matrix([[21, 17, 6], [-5, -1, -6], [4, 4, 16]])
+    expAt = Matrix([
```
Comment (mijo2) [inline sympy/solvers/tests/test_ode.py]: This ```expAt``` isn't used. Should I remove or update this?

File: sympy/solvers/tests/test_ode.py [hunk 4]
```diff
@@ -2957,8 +3158,33 @@ def test_sysode_linear_neq_order1():
                Eq(Z2(t), C4*(-k20 - k21 - k23 + k30)*exp(t*(-k20 - k21 - k23))/k23),
                Eq(Z3(t), C2*exp(-k30*t) + C4*exp(t*(-k20 - k21 - k23)))]
 
-    assert dsolve(eq, simplify=False) == sols_eq
+    assert dsolve(eq1, simplify=False) == sol1
```
Comment (mijo2) [inline sympy/solvers/tests/test_ode.py]: @oscarbenjamin Assertion error is coming for this. Should I update the ```sol1```?

File: sympy/solvers/tests/test_ode.py (hunk 3 above)
Comment (oscarbenjamin) [reply sympy/solvers/tests/test_ode.py]: I guess there should be an assert for this

Comment (oscarbenjamin) [reply sympy/solvers/tests/test_ode.py]: If you look below many things are commented. Those are things that should work and my intention was to fix all those by the time this PR got merged.

Comment (mijo2) [reply sympy/solvers/tests/test_ode.py]: Then what would be the ```A``` for this particular ```expAt```?

File: sympy/solvers/tests/test_ode.py (hunk 4 above)
Comment (mijo2) [reply sympy/solvers/tests/test_ode.py]: I have updated the ```sol1``` for now.

Comment (oscarbenjamin) [reply sympy/solvers/tests/test_ode.py]: In what way has it changed?

File: sympy/solvers/tests/test_ode.py (hunk 3 above)
Comment (oscarbenjamin) [reply sympy/solvers/tests/test_ode.py]: Presumably the one right above

File: sympy/solvers/tests/test_ode.py (hunk 4 above)
Comment (mijo2) [reply sympy/solvers/tests/test_ode.py]: @oscarbenjamin It became:

```
    sol1 = [Eq(Z0(t), C1*k10/k01 + C2*(-k10 + k30)*exp(-k30*t)/(k01 + k10 - k30) - C3*exp(t*(-k01 - k10)) + C4*(-k10*k20 - k10*k21 + k10*k30 + k20**2 + k20*k21 + k20*k23 - k20*k30 - k23*k30)*exp(t*(-k20 - k21 - k23))/(k23*(-k01 - k10 + k20 + k21 + k23))),
            Eq(Z1(t), C1 - C2*k01*exp(-k30*t)/(k01 + k10 - k30) + C3*exp(t*(-k01 - k10)) + C4*(-k01*k20 - k01*k21 + k01*k30 + k20*k21 + k21**2 + k21*k23 - k21*k30)*exp(t*(-k20 - k21 - k23))/(k23*(-k01 - k10 + k20 + k21 + k23))),
            Eq(Z2(t), C4*(-k20 - k21 - k23 + k30)*exp(t*(-k20 - k21 - k23))/k23),
            Eq(Z3(t), C2*exp(-k30*t) + C4*exp(t*(-k20 - k21 - k23)))]
```
From the ```sol1``` mentioned above.

File: sympy/solvers/tests/test_ode.py (hunk 3 above)
Comment (mijo2) [reply sympy/solvers/tests/test_ode.py]: It is not the case actually. The one just below corresponds with the A mentioned in the code.

File: sympy/solvers/tests/test_ode.py (hunk 4 above)
Comment (oscarbenjamin) [reply sympy/solvers/tests/test_ode.py]: Does that mean that the commented checksysodesol below works now? (It should be uncommented if so)

File: sympy/solvers/tests/test_ode.py (hunk 3 above)
Comment (oscarbenjamin) [reply sympy/solvers/tests/test_ode.py]: We can work backwards using log:
```
In [70]:     expAt = Matrix([ 
    ...:         [(8*t*exp(12*t) + 5*exp(12*t) - 1)*exp(4*t)/4, 
    ...:          (8*t*exp(12*t) + 5*exp(12*t) - 5)*exp(4*t)/4, 
    ...:          (exp(12*t) - 1)*exp(4*t)/2], 
    ...:         [(-8*t*exp(12*t) - exp(12*t) + 1)*exp(4*t)/4, 
    ...:          (-8*t*exp(12*t) - exp(12*t) + 5)*exp(4*t)/4, 
    ...:          (-exp(12*t) + 1)*exp(4*t)/2], 
    ...:         [4*t*exp(16*t), 4*t*exp(16*t), exp(16*t)]])                                                                                   

In [71]: expAt.log()                                                                                                                           
Out[71]: 
⎡         ⎛ 4⋅t⎞        ⎛ 16⋅t⎞             ⎛ 4⋅t⎞        ⎛ 16⋅t⎞       ⎛ 4⋅t⎞      ⎛ 16⋅t⎞⎤
⎢      log⎝ℯ   ⎠   5⋅log⎝ℯ    ⎠        5⋅log⎝ℯ   ⎠   5⋅log⎝ℯ    ⎠    log⎝ℯ   ⎠   log⎝ℯ    ⎠⎥
⎢2⋅t - ───────── + ────────────  2⋅t - ─────────── + ────────────  - ───────── + ──────────⎥
⎢          4            4                   4             4              2           2     ⎥
⎢                                                                                          ⎥
⎢          ⎛ 4⋅t⎞      ⎛ 16⋅t⎞               ⎛ 4⋅t⎞      ⎛ 16⋅t⎞       ⎛ 4⋅t⎞      ⎛ 16⋅t⎞ ⎥
⎢       log⎝ℯ   ⎠   log⎝ℯ    ⎠          5⋅log⎝ℯ   ⎠   log⎝ℯ    ⎠    log⎝ℯ   ⎠   log⎝ℯ    ⎠ ⎥
⎢-2⋅t + ───────── - ──────────   -2⋅t + ─────────── - ──────────    ───────── - ────────── ⎥
⎢           4           4                    4            4             2           2      ⎥
⎢                                                                                          ⎥
⎢                                                                            ⎛ 16⋅t⎞       ⎥
⎣             4⋅t                              4⋅t                        log⎝ℯ    ⎠       ⎦

In [72]: tp = Symbol('t', positive=True)                                                                                                       

In [73]: expAt.log().subs(t, tp)/tp                                                                                                            
Out[73]: 
⎡21  17  6 ⎤
⎢          ⎥
⎢-5  -1  -6⎥
⎢          ⎥
⎣4   4   16⎦
```

Comment (oscarbenjamin) [reply sympy/solvers/tests/test_ode.py]: We can work backwards using log:
```
In [70]:     expAt = Matrix([ 
    ...:         [(8*t*exp(12*t) + 5*exp(12*t) - 1)*exp(4*t)/4, 
    ...:          (8*t*exp(12*t) + 5*exp(12*t) - 5)*exp(4*t)/4, 
    ...:          (exp(12*t) - 1)*exp(4*t)/2], 
    ...:         [(-8*t*exp(12*t) - exp(12*t) + 1)*exp(4*t)/4, 
    ...:          (-8*t*exp(12*t) - exp(12*t) + 5)*exp(4*t)/4, 
    ...:          (-exp(12*t) + 1)*exp(4*t)/2], 
    ...:         [4*t*exp(16*t), 4*t*exp(16*t), exp(16*t)]])                                                                                   

In [71]: expAt.log()                                                                                                                           
Out[71]: 
⎡         ⎛ 4⋅t⎞        ⎛ 16⋅t⎞             ⎛ 4⋅t⎞        ⎛ 16⋅t⎞       ⎛ 4⋅t⎞      ⎛ 16⋅t⎞⎤
⎢      log⎝ℯ   ⎠   5⋅log⎝ℯ    ⎠        5⋅log⎝ℯ   ⎠   5⋅log⎝ℯ    ⎠    log⎝ℯ   ⎠   log⎝ℯ    ⎠⎥
⎢2⋅t - ───────── + ────────────  2⋅t - ─────────── + ────────────  - ───────── + ──────────⎥
⎢          4            4                   4             4              2           2     ⎥
⎢                                                                                          ⎥
⎢          ⎛ 4⋅t⎞      ⎛ 16⋅t⎞               ⎛ 4⋅t⎞      ⎛ 16⋅t⎞       ⎛ 4⋅t⎞      ⎛ 16⋅t⎞ ⎥
⎢       log⎝ℯ   ⎠   log⎝ℯ    ⎠          5⋅log⎝ℯ   ⎠   log⎝ℯ    ⎠    log⎝ℯ   ⎠   log⎝ℯ    ⎠ ⎥
⎢-2⋅t + ───────── - ──────────   -2⋅t + ─────────── - ──────────    ───────── - ────────── ⎥
⎢           4           4                    4            4             2           2      ⎥
⎢                                                                                          ⎥
⎢                                                                            ⎛ 16⋅t⎞       ⎥
⎣             4⋅t                              4⋅t                        log⎝ℯ    ⎠       ⎦

In [72]: tp = Symbol('t', positive=True)                                                                                                       

In [73]: expAt.log().subs(t, tp)/tp                                                                                                            
Out[73]: 
⎡21  17  6 ⎤
⎢          ⎥
⎢-5  -1  -6⎥
⎢          ⎥
⎣4   4   16⎦
```

**Comment by czgdp1807** [comment]
Closing in favour of https://github.com/sympy/sympy/pull/18720

**Pull request closed without merging**